src/file.py

```diff
@@ -0,0 +1,338 @@
+        req = cs3sp.TouchFileRequest(
+            ref=resource.ref,
+            opaque=types.Opaque(map={"Upload-Length": types.OpaqueEntry(decoder="plain", value=str.encode("0"))}),
+        )
+        res = self._gateway.TouchFile(request=req, metadata=[self._auth.get_token()])
+        if res.status.code == cs3code.CODE_FAILED_PRECONDITION:
+            self._log_precondition_info(resource, res, "touch")
+            raise FileLockedException()
+        if res.status.code == cs3code.CODE_UNAUTHENTICATED:
+            self._log_authentication_error(resource, res, "touch")
+            raise AuthenticationException()
+        if res.status.code != cs3code.CODE_OK:
+            self._log_unknown_error(resource, res, "touch")
+            raise IOError(res.status.message)
+        self._log.debug(f'msg="Invoked touchfile" result="{res}"')
+
+    def write_file(self, resource: Resource, content: str | bytes, size: int, islock=False) -> None:
+        """Write a file using the given userid as access token. The entire content is written
+        and any pre-existing file is deleted (or moved to the previous version if supported).
+        The islock flag is currently not supported. The backend should at least support
+        writing the file with O_CREAT|O_EXCL flags to prevent races."""
+        tstart = time.time()
+        # prepare endpoint
+        if size == -1:
+            if isinstance(content, str):
+                content = bytes(content, "UTF-8")
+            size = len(content)
+        req = cs3sp.InitiateFileUploadRequest(
+            ref=resource.ref,
+            opaque=types.Opaque(map={"Upload-Length": types.OpaqueEntry(decoder="plain", value=str.encode(str(size)))}),
+        )
+        res = self._gateway.InitiateFileUpload(request=req, metadata=[self._auth.get_token()])
+        if res.status.code == cs3code.CODE_FAILED_PRECONDITION:
+            self._log_precondition_info(resource, res, "InitiateFileUpload")
+            raise FileLockedException()
+        if res.status.code == cs3code.CODE_UNAUTHENTICATED:
+            self._log_authentication_error(resource, res, "InitiateFileUpload")
+            raise AuthenticationException()
+        if res.status.code != cs3code.CODE_OK:
+            self._log_unknown_error(resource, res, "InitiateFileUpload")
+            raise IOError(res.status.message)
+        tend = time.time()
+        self._log.debug(
+            f'msg="writefile: InitiateFileUploadRes returned" trace="{res.status.trace}" protocols="{res.protocols}"'
+        )
+
+        # Upload
+        try:
+            protocol = [p for p in res.protocols if p.protocol in ["simple", "spaces"]][0]
+            if self._config.tus_enabled:
+                headers = {
+                    "Tus-Resumable": "1.0.0",
+                    "File-Path": resource.file,
+                    "File-Size": str(size),
+                    "X-Access-Token": self._auth.get_token()[1],
```

Comment:
Don't like this that much. Would this look better and not be overly costly?

```suggestion
                    **dict([self._auth.get_token()]),
```

Reply:
If you do change, please change in the other places as well.

Reply:
Me and Giuseppe discussed it and decided that **dict([self._auth.get_token()]) is the best option